pkg/vm/engine/tae/txn/txnbase/logtail.go

```diff
@@ -0,0 +1,234 @@
+			txns:  txns,
+		}
+		l.pages = append(l.pages, newPage)
+	}
+}
+
+func (l *LogtailMgr) SyncLogtails(req *api.SyncLogTailReq) (api.SyncLogTailResp, error) {
+	if tid := req.Table.TbId; tid == constSysTblDBID || tid == constSysTblTblID {
+		return l.syncCatalogLogtails(req)
+	}
+	return l.syncDataLogtails(req)
+}
+
+func (l *LogtailMgr) syncDataLogtails(req *api.SyncLogTailReq) (api.SyncLogTailResp, error) {
+	start := types.BuildTS(req.CnHave.PhysicalTime, req.CnHave.LogicalTime)
+	end := types.BuildTS(req.CnWant.PhysicalTime, req.CnWant.LogicalTime)
+	_ /* dirtyBlks */ = l.CollectDirtyTree(start, end, req.Table.TbId)
+	// TODO: return scanDataLogtail(dbID, tableID, dirtyBlks)
+	return api.SyncLogTailResp{}, nil
+}
+
+func (l *LogtailMgr) syncCatalogLogtails(req *api.SyncLogTailReq) (api.SyncLogTailResp, error) {
+	start := types.BuildTS(req.CnHave.PhysicalTime, req.CnHave.LogicalTime)
+	end := types.BuildTS(req.CnWant.PhysicalTime, req.CnWant.LogicalTime)
+	_ /* changed */ = l.CollectCatalogChange(start, end)
+	// TODO: return scanCatalogLogtail()
+	return api.SyncLogTailResp{}, nil
+}
+
+// scanTxnBetween happens concurrently, interleaves with AddTxn
+func (l *LogtailMgr) scanTxnBetween(start, end types.TS, f func(txn txnif.AsyncTxn) (moveOn bool)) {
+	l.rwlock.RLock()
+	defer l.rwlock.RUnlock()
+
+	// i is start page idx, where the start ts lands in
+	i := bisect_ref(l.pages, func(page *txnPage) bool { return page.maxTs.GreaterEq(start) })
+	if i < len(l.pages) {
+		txns := l.pages[i].txns
+		// j is the first txn whose prepareTS >= start ts
+		// no need to check j outbound because i is determined
+		j := bisect(txns, func(txn txnif.AsyncTxn) bool { return txn.GetPrepareTS().GreaterEq(start) })
+		for _, txn := range txns[j:] {
+			if txn.GetPrepareTS().Greater(end) {
+				return
+			}
+			if !f(txn) {
+				return
+			}
+		}
+
+	}
+	// scan other pages
+	if i+1 < len(l.pages) {
+		for _, page := range l.pages[i+1:] {
+			for _, txn := range page.txns {
+				if txn.GetPrepareTS().Greater(end) {
+					return
+				}
+				if !f(txn) {
+					return
+				}
+			}
+		}
+	}
+
+	// end ts is not found in pages, has to check active page
+	// getting active page readonly view will not block AddTxn
+	size := atomic.LoadInt32(l.activeSize)
+	view := l.activePage[:size]
+	viewStart := 0
+	if len(l.pages) == 0 {
+		viewStart = bisect(view, func(txn txnif.AsyncTxn) bool { return txn.GetPrepareTS().GreaterEq(start) })
+		if viewStart >= int(size) { // start ts is too big
+			return
+		}
+	}
+	for _, txn := range view[viewStart:] {
+		if txn.GetPrepareTS().Greater(end) {
+			return
+		}
+		if !f(txn) {
+			return
+		}
+	}
+}
+
+func (l *LogtailMgr) CollectDirtyTree(start, end types.TS, tableID uint64) map[uint64]map[uint64]struct{} {
+	// if tree[segID] is nil, it means that is just a seg operation
+	tree := make(map[uint64]map[uint64]struct{}, 0)
+	var blkSet map[uint64]struct{}
+	var exist bool
+	f := func(txn txnif.AsyncTxn) (moveOn bool) {
```

Comment:
When scanning a page, maybe it needs to retain a read lock. Do some simple data processing during the scan process, and put complex and time-consuming data processing after the scan.